cli/flox-rust-sdk/src/providers/catalog.rs

```diff
@@ -50,48 +30,70 @@ impl CatalogClient {
             client: APIClient::new(DEFAULT_CATALOG_URL),
         }
     }
+
+    pub fn new_with_reqwest_client(client: reqwest::Client) -> Self {
+        Self {
+            client: APIClient::new_with_client(DEFAULT_CATALOG_URL, client),
+        }
+    }
 }
+
+#[derive(Debug)]
+pub struct MockClient;
+
 impl Default for CatalogClient {
     fn default() -> Self {
         Self::new()
     }
 }
 
-#[derive(Debug)]
-pub struct MockClient;
-
-impl Client {
+#[async_trait]
+#[enum_dispatch]
+pub trait ClientTrait {
     /// Resolve a list of [PackageGroup]s into a list of
     /// [ResolvedPackageGroup]s.
-    pub async fn resolve(
+    async fn resolve(
+        &self,
+        package_groups: Vec<PackageGroup>,
+    ) -> Result<Vec<ResolvedPackageGroup>, CatalogClientError>;
+}
+
+#[async_trait]
+impl ClientTrait for CatalogClient {
+    async fn resolve(
         &self,
         package_groups: Vec<PackageGroup>,
     ) -> Result<Vec<ResolvedPackageGroup>, CatalogClientError> {
-        match self {
-            Client::Catalog(client) => {
-                let package_groups = api_types::PackageGroups {
-                    items: package_groups
-                        .into_iter()
-                        .map(TryInto::try_into)
-                        .collect::<Result<Vec<_>, _>>()?,
-                };
-
-                let response = client
-                    .client
-                    .resolve_api_v1_catalog_resolve_post(&package_groups)
-                    .await
-                    .map_err(CatalogClientError::Resolution)?;
-
-                let resolved_package_groups = response.into_inner();
-
-                resolved_package_groups
-                    .items
-                    .into_iter()
-                    .map(TryInto::try_into)
-                    .collect::<Result<Vec<_>, _>>()
-            },
-            Client::Mock(_) => unimplemented!(),
-        }
+        let package_groups = api_types::PackageGroups {
+            items: package_groups
+                .into_iter()
+                .map(TryInto::try_into)
+                .collect::<Result<Vec<_>, _>>()?,
+        };
+
+        let response = self
+            .client
+            .resolve_api_v1_catalog_resolve_post(&package_groups)
+            .await
+            .map_err(CatalogClientError::Resolution)?;
+
```

Comment:
With exposing an explicit async interface we risk infecting larege parts of flox with `async` code once again.
If we expose a sync interface that is just trying to `block_on` the client future, the fact that we're running as w whole within a tokio runtime means that trying to call `block_on` will panic.
Running this future in a different executor (such as e.g. `futures::executor`) allegedly causes deadlocks on single core systems (which some docker containers might be)
Thus, short of removing the "parent" async runtime or bubbling through a psecial runtime for clients, best we can do is do the async call in a separate thread with a "fresh" runtime.

ping @mkenigs @zmitchell to discuss


```suggestion
        let response = std::thread::scope(|s| {
            s.spawn(|| {
                let rt = tokio::runtime::Runtime::new().unwrap();
                rt.block_on(
                    self.client
                        .resolve_api_v1_catalog_resolve_post(&package_groups),
                )
            })
            .join()
        })
        .unwrap()
        .map_err(CatalogClientError::Resolution)?;
```

Reply:
+1 for KISS, which usually implies synchronous / blocking. I was surprised to see as much async as there is - the job we're doing is inherently blocking from the user perspective, and there's not a lot for us to be doing in parallel either now or in future when we get on to builds. Would love to talk over all our current and planned uses of async at some point.